File: transcendence/data/django/project/hook.py
Table = {}

def Add( name, id, func ):
	if (not name in Table):
		Table[ name ] = {}
	Table[ name ][ id ] = func

def Remove( name, id ):
	if ( not name in Table ):
		return
	del Table[ name ][ id ]

def Call( name, args ):
	if ( not name in Table ):
		return
	for id in Table[ name ]:
		Table[ name ][ id ]( args )
	

File: transcendence/data/django/app_index/static/index/js/hook.js
hook = {}
hook.table = {}

hook.Add = function( name, id, func )
{
	if ( !hook.table[ name ] ){
		hook.table[ name ] = {}
	}

	hook.table[ name ][ id ] = func
}

hook.Remove = function( name, id )
{
	if ( !hook.table[ name ] ){ return; }
	
	delete hook.table[ name ][ id ];
}

hook.Call = function( name, args )
{
	if ( !hook.table[ name ] ){ return; }

	const entries = Object.entries( hook.table[ name ] );
	entries.forEach( function( [key, value] ) {
		value( args )
	});
}